crates/prune/src/pruner.rs

```diff
@@ -119,37 +97,16 @@ impl<DB: Database> Pruner<DB> {
                 .map_or(1, |previous_tip_block_number| tip_block_number - previous_tip_block_number)
                 as usize;
 
-        // TODO(alexey): this is cursed, refactor
-        let segments: [PrunableSegment<DB>; 5] = [
-            PrunableSegment::new(
-                segments::Receipts::default(),
-                PruneModes::prune_target_block_receipts,
-            ),
-            PrunableSegment::new(
-                segments::TransactionLookup::default(),
-                PruneModes::prune_target_block_transaction_lookup,
-            ),
-            PrunableSegment::new(
-                segments::SenderRecovery::default(),
-                PruneModes::prune_target_block_sender_recovery,
-            ),
-            PrunableSegment::new(
-                segments::AccountHistory::default(),
-                PruneModes::prune_target_block_account_history,
-            ),
-            PrunableSegment::new(
-                segments::StorageHistory::default(),
-                PruneModes::prune_target_block_storage_history,
-            ),
-        ];
-
-        for PrunableSegment(segment, get_prune_target_block) in segments {
+        for segment in &self.segments {
             if delete_limit == 0 {
                 break
             }
 
-            if let Some((to_block, prune_mode)) =
-                get_prune_target_block(&self.modes, tip_block_number)?
+            if let Some((to_block, prune_mode)) = segment
+                .mode()
+                .map(|mode| mode.prune_target_block(tip_block_number, segment.segment()))
+                .transpose()?
+                .flatten()
```

Comment:
this could be moved to a default trait function?
but perhaps not worth it if we only call this once?

Reply:
didn't want to move to a default trait method what shouldn't be changed by the underlying implementation. When it works that way as it is now, we're sure how it's called, rather than relying on every segment not re-implementing the default trait method.